src/app/registration/register.service.spec.ts

```diff
@@ -2,18 +2,61 @@ import { TestBed, inject } from '@angular/core/testing';
 
 import { RegisterService } from './register.service';
 import { ApiService } from '../_services/api.service';
+import { UserRegistration } from '../user/user-registration';
+import { User, Role } from '../user/user';
+
+const uniqueUsername = 'uniqueUsername';
+const duplicateUsername = 'duplicateUsername';
+const uniqueEmail = 'uniqueEmail@junk.com';
+const duplicateEmail = 'duplicateEmail@junk.com';
+const userRegistration = new UserRegistration('abc123', uniqueUsername, 2, uniqueEmail, 'Test Success Fullname');
+const role = new Role(2, 'userRole 2 definition');
+const user = new User(20, role, 'abc123', uniqueUsername, 1, uniqueEmail, 'Test Fullname', true, 'demograhic indicator' );
+const userBadRegistration = new UserRegistration('abc123', duplicateUsername, 2, duplicateEmail, 'Test Failed Fullname');
+const failedUserReg = '{"errors": [{"field": "email","message": "email.not.unique"}, {"field": "name", "message": "name.not.unique"}]}';
 
 describe('RegisterService', () => {
   beforeEach(() => {
     TestBed.configureTestingModule({
       providers: [
-      	{ provide: RegisterService, useClass: RegisterService },
-		    { provide: ApiService, useValue: { post: jasmine.createSpy('post') }},
+        { provide: RegisterService, useClass: RegisterService },
+        { provide: ApiService, useValue: { post: jasmine.createSpy('post') }},
        ]
     });
   });
 
   it('should be created', inject([RegisterService], (service: RegisterService) => {
     expect(service).toBeTruthy();
   }));
+
+  it('isUniqueUsername(uniqueUsername) should be truthy', inject([RegisterService], (service: RegisterService) => {
+    spyOn(service, 'isUniqueUsername').and.returnValue(Promise.resolve(true));
```

Comment:
First, very nice to see you using Spies and all that good stuff.. This is where you begin to get into the meat of unit testing..

What you want to do with this tests, is verify that a method operates as you would expect. So, for the method isUniqueUsername(), we are able to follow the following steps in its code:

- Well, there's only one, this is a wrapper for an API call, who's details are hidden to us. Our One Step, is to return the value from this API.

The point of a Spy is to allow us to define what happens when a method is called. This is useful because when we are testing, we can create our own conditions and verify what happens when a method is called.

In this case, we want to verify what happens when the API is called. So, we set our Spy, not on the method we are testing, but on the method who's behavior we want to control.

```
it('is a test', inject([RegisterService, ApiService], (service: RegisterService, apiService: ApiService) => {
    spyOn(apiService, 'post').and.returnValue(Promise.resolve(true));
...
}
```

And then, as you have it already, we would check to see if the method we are testing, `service.isUniqueUsername` returns true-ish, as we would expect.

But what you have will not work, because `apiService` is returning a Promise; it is asynchronous, and the code you've written assumes it is not. See this answer on StackOverflow: https://stackoverflow.com/a/35431221/512561

Reply:
Hi Johnathan (@haxwell),

I revised the register service tests on the devRegTests branch per your guidance above. Yet, I'm still not sure I have the revised tests correctly. Also,  what should afterEach() do to remove the injected services. Or can afterEach() be eliminated for these tests.

Feedback would be appreciated. 

Deborah